enclave/core/sgx/td.c

```diff
@@ -163,8 +159,28 @@ void* td_to_tcs(const oe_sgx_td_t* td)
 oe_sgx_td_t* oe_sgx_get_td()
 {
     oe_sgx_td_t* td;
-
-    asm("mov %%fs:0, %0" : "=r"(td));
+    void* fsbase;
+    void* gsbase;
+
+    asm("mov %%fs:0, %0" : "=r"(fsbase));
+    asm("mov %%gs:0, %0" : "=r"(gsbase));
+
+    td = (oe_sgx_td_t*)fsbase;
+
+    if (fsbase != gsbase)
+    {
+        /* The mismatch between FS and GS indicates that FS could have
+         * been changed by the application, which should be responsible for
+         * managing the FS values. Abort in such cases if the enclave is not in
+         * simulation mode (on Windows, GS could also be changed). */
+        if (td_initialized(td) && !td->simulate)
+        {
+            /* Cannot call oe_abort here as the call path also uses
+             * oe_sgx_get_td. Set the state to ABORTED instead such
+             * that the following enclave entries are blocked */
+            td->state = OE_TD_STATE_ABORTED;
```

Comment:
Should we raise an exception here..say a SEGV so that the execution does not continue/the programmer is aware?